File: shared/bw_shared/pid/pid.py
from bw_shared.pid.config import PidConfig


class PID:
    """
    Implements a PID controller with optional feedforward term and integral zone.
    """

    def __init__(self, config: PidConfig):
        """
        Config arguments:
        kp: Proportional gain.
        ki: Integral gain.
        kd: Derivative gain.
        kf: Feedforward gain.
        i_zone: Integral zone. If the error is within this zone, the integral term is
            accumulated. If not set, the integral term is always accumulated.
        i_max: Maximum value of the integral term. If not set, the integral term is not
            limited.
        tolerance: Error tolerance for which the controller is considered to have reached the
            setpoint. Output will be set to zero if the error is within this tolerance.
        """
        self.kp = config.kp
        self.ki = config.ki
        self.kd = config.kd
        self.kf = config.kf
        self.i_zone = config.i_zone
        self.i_max = config.i_max
        self.tolerance = config.tolerance
        self.i_accum = 0.0
        self.prev_error = 0.0

    def reset(self) -> None:
        """
        Reset the integral term and previous error.
        """
        self.i_accum = 0.0
        self.prev_error = 0.0

    def update(self, setpoint: float, measurement: float, dt: float) -> float:
        """
        Update the PID controller with the current setpoint and measurement.

        Args:
            setpoint: Desired value.
            measurement: Current value.
            dt: Time step.

        Returns:
            Output value.
        """
        error = setpoint - measurement
        if abs(error) < self.tolerance:
            return 0.0
        output = 0.0
        output += self._calculate_p(error)
        output += self._calculate_i(error)
        output += self._calculate_d(error, dt)
        output += self._calculate_f(setpoint)
        return output

    def _calculate_p(self, error: float) -> float:
        if self.kp == 0.0:
            return 0.0
        return self.kp * error

    def _calculate_i(self, error: float) -> float:
        if self.ki == 0.0:
            return 0.0
        if self.i_zone is None:
            self.i_accum += error
        elif abs(error) < self.i_zone:
            self.i_accum += error

        if self.i_max != 0.0:
            if self.i_accum > 0.0:
                self.i_accum = min(self.i_accum, self.i_max / self.ki)
            else:
                self.i_accum = max(self.i_accum, -self.i_max / self.ki)

        return self.ki * self.i_accum

    def _calculate_d(self, error: float, dt: float) -> float:
        if self.kd == 0.0:
            return 0.0
        if dt < 0.0:
            return 0.0
        output = self.kd * (error - self.prev_error) / dt
        self.prev_error = error
        return output

    def _calculate_f(self, setpoint: float) -> float:
        if self.kf == 0.0:
            return 0.0
        return self.kf * setpoint

File: simulation/TrueBattleBotSim/Assets/Scripts/RobotControl/PID.cs
using UnityEngine;

public class PID
{
    private float kp, ki = 0.0f, kd = 0.0f, kf = 0.0f;
    private float i_zone = 0.0f, i_max = 0.0f;
    private float epsilon = 1e-9f, tolerance = 0.0f;
    private float i_accum = 0.0f;
    private float prev_error = float.NaN;

    public PID(PidConfig config)
    {
        SetConfig(config);
    }

    public void SetConfig(PidConfig config)
    {
        kp = config.kp;
        ki = config.ki;
        kd = config.kd;
        kf = config.kf;
        i_zone = config.i_zone;
        i_max = config.i_max;
        epsilon = config.epsilon;
        tolerance = config.tolerance;

        Reset();
    }

    public void Reset()
    {
        i_accum = 0.0f;
        prev_error = float.NaN;
    }

    public float Update(float setpoint, float measurement, float dt)
    {
        float error = setpoint - measurement;
        if (Mathf.Abs(error) < tolerance)
        {
            return 0.0f;
        }
        if (dt <= 0.0f)
        {
            Debug.LogWarning("dt must be positive");
            return 0.0f;
        }
        float output = 0.0f;
        output += CalculateP(error);
        output += CalculateI(error, dt);
        output += CalculateD(error, dt);
        output += CalculateF(setpoint);
        return output;
    }

    private float CalculateP(float error)
    {
        if (Mathf.Abs(kp) < epsilon)
        {
            return 0.0f;
        }
        return kp * error;
    }
    private float CalculateI(float error, float dt)
    {
        if (Mathf.Abs(ki) < epsilon)
            return 0.0f;
        if (i_zone <= epsilon || Mathf.Abs(error) < i_zone)
            i_accum += error;
        else
            i_accum = 0.0f;

        if (i_max > epsilon)
            if (i_accum > 0.0f)
                i_accum = Mathf.Min(i_accum, i_max / ki);
            else
                i_accum = Mathf.Max(i_accum, -i_max / ki);

        return ki * i_accum * dt;
    }

    private float CalculateD(float error, float dt)
    {
        if (Mathf.Abs(kd) < epsilon)
            return 0.0f;
        if (float.IsNaN(prev_error))
            prev_error = error;
        float derivative = (error - prev_error) / dt;
        prev_error = error;
        return kd * derivative;
    }

    private float CalculateF(float setpoint)
    {
        if (Mathf.Abs(kf) < epsilon)
            return 0.0f;
        return kf * setpoint;
    }
}
